haystack/tools/from_function.py

```diff
@@ -109,15 +124,33 @@ def get_weather(
         if param_name in schema["properties"]:
             schema["properties"][param_name]["description"] = param_description
 
-    return Tool(name=name or function.__name__, description=tool_description, parameters=schema, function=function)
-
-
-def tool(function: Callable) -> Tool:
+    return Tool(
+        name=name or function.__name__,
+        description=tool_description,
+        parameters=schema,
+        function=function,
+        inputs_from_state=inputs,
+        outputs_to_state=outputs,
+    )
+
+
+def tool(
+    function: Optional[Callable] = None,
```

Comment:
why `function` is `Optional`?

Reply:
My understanding is that this is necessary because we want to support `tool` as decorator both with and without arguments. Without arguments, the decorator returns a decorated version of the function that is decorated. With arguments, it returns a decorator that beforehand processes the provided input arguments, such as name or description.
Only if `function` is typed as `Optional` we can check for `if function is None:` so that we can then return a new decorator function that sets the name, description etc. to the customized values. Otherwise we decorate the function directly, which gives no option for customization. For example, if you want to decorate a function with `@tool` but also want to change the description.

Reply:
Thanks for the explanation.
I tried it and works as expected.